File: Leetcode Solutions/2 Weeks Study Plan to Tackle DS/DS 14-Day Challenge/Day-14/Two Sum IV - Input is a BST.java
class Solution {
    public boolean findTarget(TreeNode root, int k) {
       List<Integer> nums = new ArrayList<>();
       inorder(root, nums);
       for(int i = 0, j = nums.size()-1; i<j;){
           if(nums.get(i) + nums.get(j) == k)return true;
           if(nums.get(i) + nums.get(j) < k)i++;
           else j--;
       }
       return false;
   }
   public void inorder(TreeNode root, List<Integer> nums){
       if(root == null)return;
       inorder(root.left, nums);
       nums.add(root.val);
       inorder(root.right, nums);
   }
}

File: Leetcode Solutions/LeetCode October Challenge(2022)/Two Sum IV - Input is a BST.cpp
class Solution {
public:
    bool findTarget(TreeNode* root, int target) {
   if(!root) return false;
    stack<TreeNode*> s1, s2;
    TreeNode* temp1 = root;
    TreeNode* temp2 = root;
    while(1)
    {
        while(temp1)
        {
            s1.push(temp1);
            temp1 = temp1->left;
        }
        while(temp2)
        {
            s2.push(temp2);
            temp2 = temp2->right;
        }
        if(s1.empty() || s2.empty()) break;
        temp1 = s1.top();
        temp2 = s2.top();
        if((temp1->val + temp2->val) == target)
        {
            if(temp1 == temp2) return false;
            return true;
        }
        else if((temp1->val + temp2->val) < target)
        {
            s1.pop();
            temp1 = temp1->right;
            temp2 = NULL;
        }
        else
        {
            s2.pop();
            temp2 = temp2->left;
            temp1 = NULL;
        }
    }
    return false;
}
};